Automattic/wp-calypso
# Checkout: Centralize logic that determines which contact form to use

#### Changes proposed in this Pull Request

This converts the `WPContactForm` and its dependent components into TypeScript, and fixes some bugs along the way.

Notably, the use of booleans to determine if we should display the domains contact form, the G Suite contact form, or the tax contact form has caused numerous bugs. This PR creates a single function, `getContactDetailsType` which will examine the current cart and return a string that identifies which contact details form should be displayed. This centralizes the logic and makes it easier to prevent edge cases where the wrong form is displayed.

The intentional behavior changes caused by this are highlighted in the testing instructions below.

Domain contact form:

<img width="584" alt="Screen Shot 2020-11-04 at 6 23 36 PM" src="https://user-images.githubusercontent.com/2036909/98178754-46cd8300-1ecb-11eb-9a7c-90f9992abe05.png">

Taxes contact form:

<img width="574" alt="Screen Shot 2020-11-04 at 6 26 19 PM" src="https://user-images.githubusercontent.com/2036909/98178766-4cc36400-1ecb-11eb-87fc-e1c8e162c013.png">

G Suite contact form:

<img width="575" alt="Screen Shot 2020-11-04 at 6 31 15 PM" src="https://user-images.githubusercontent.com/2036909/98179091-facf0e00-1ecb-11eb-889b-4d224d359d66.png">

Fixes https://github.com/Automattic/wp-calypso/issues/45301 
Fixes https://github.com/Automattic/wp-calypso/issues/45786

#### Testing instructions

1. Visit checkout with just a plan in the cart. Verify that you see the tax contact form (just postal code and country).
1. Visit checkout with just a plan in the cart and enough credits to fully cover the purchase. Verify that you see no contact form. **This is new behavior**
1. Visit checkout with just a plan in the cart and a 100% discount coupon. Verify that you see no contact form.
1. Visit checkout with a domain registration and a plan in the cart. Verify that you see the domain contact form (name, address, etc.)
1. Visit checkout with a domain registration and a plan in the cart and enough credits to fully cover the purchase. Verify that you see the domain contact form (name, address, etc.)
1. Visit checkout with a domain registration and a plan in the cart and a 100% discount coupon. Verify that you see the domain contact form (name, address, etc.)
1. Visit checkout with just a domain transfer in the cart. Verify that you see the domain contact form (name, address, etc.)
1. Visit checkout with just a domain mapping in the cart. Verify that you see the domain contact form (name, address, etc.)
1. Visit checkout with a domain registration and a new G Suite subscription in the cart. Verify that you see the domain contact form (name, address, etc.)
1. Visit checkout with just a new G Suite subscription in the cart. Verify that you see the G Suite contact form (name, alternate email address, postal code and country).
1. Visit checkout with extra G Suite licenses in the cart. Verify that you see the tax contact form (just postal code and country). **This is new behavior**
